File: Jacobi.f90
! Anant Shah
! Jacobi Rotation
! Date : 27-12-2017

	program JacobiRotation

        implicit none

        integer :: dim_mat ! Matrix dimension
        real(kind = 16),dimension(:,:),allocatable               :: mat ! Cauchy matrix 
        real(kind = 16),dimension(:,:),allocatable               :: mat_rot ! Roation matrix
        real(kind = 16),dimension(:),allocatable                 ::eigenvalues
        real :: square_norm
        real :: diag_square_norm
        real :: offdiag_square_norm
        integer :: i,j ! Loop control variables
        real,parameter  :: limit = 1.0e-10
        integer :: rot_count

        print *,"Enter the dimension of the matrix : "
        read *,dim_mat ! Read the dimension of the matrix; users input

        ! Allocate memory to the rotation matrix and the Cauchy matrix
        allocate(mat(dim_mat,dim_mat),mat_rot(dim_mat,dim_mat) )
        allocate( eigenvalues(dim_mat) )

        mat = mat_gen(dim_mat)

        square_norm = mat_sqr_norm(mat,dim_mat)
        print *,"Square =",square_norm
        do rot_count = 1,10
                mat_rot = jacobi(mat,dim_mat)
                mat = rotate(mat,mat_rot,dim_mat)

                ! Printing the rotation matrix
                print *,mat_rot

                diag_square_norm = mat_diag_sqr_norm(mat,dim_mat)
                offdiag_square_norm = mat_offdiag_sqr_norm(mat,dim_mat)
                if( offdiag_square_norm .LT. limit ) then
                        print *,"Diagoanl square =",diag_square_norm
                        print *,"Off Diagonal Square =",offdiag_square_norm
                        print *,mat
                        print *,'Exiting the program'
                        exit
                else
                        print *,"Diagonal Square =",diag_square_norm
                        print *,"Off diagonal Square =",offdiag_square_norm
                        print *,mat
                end if
        end do

        do i=1,dim_mat
                eigenvalues(i) = mat(i,i)
        end do
        print *,"Eigenvalues : "
        print *,eigenvalues

        contains
        function        mat_gen(dim_mat) result(mat)

                integer                    :: i,j ! Loop control variables
                real(kind = 16),dimension(:,:),allocatable      :: mat
                integer ,intent(in)        :: dim_mat ! Matrix dimension                

                ! Allocate the memory
                allocate( mat(dim_mat,dim_mat) )
                outer : do i=1,dim_mat
                        inner : do j=1,dim_mat
                                mat(i,j) = 1.0/( i+j )
                        end do inner
                end do outer
        end function    mat_gen

        real*8 function        mat_sqr_norm(mat,dim_mat) result(elements_sum)
                integer                 :: i,j ! Loop control variables
                real(kind = 16),dimension(:,:),allocatable    :: mat ! Cauchy matrix input
                integer    :: dim_mat ! Dimension
                elements_sum = 0.0

                outer1   : do i=1,dim_mat
                        inner1 : do j=1,dim_mat
                                elements_sum = elements_sum + mat(i,j)**2
                        end do inner1
                end do outer1
        end function    mat_sqr_norm

        real*8 function        mat_diag_sqr_norm(mat,dim_mat) result(dia_sum)
                integer         :: i,j ! Loop control variables
                real(kind = 16),dimension(:,:),allocatable   :: mat ! Cauchy matrix input
                integer,intent(in)      :: dim_mat

                dia_sum = 0.0
                do i=1,dim_mat
                        dia_sum = dia_sum + mat(i,i)**2
                end do
        end function mat_diag_sqr_norm

        real*8 function        mat_offdiag_sqr_norm(mat,dim_mat) result(off_sum)
                integer                 :: i,j ! Loop control variables
                real(kind = 16),dimension(:,:),allocatable   :: mat ! Cauchy matrix
                integer,intent(in)      :: dim_mat ! Dimension of the matrix

                off_sum = 0.0

                outer2 : do i=1,dim_mat
                        inner2 : do j=1,dim_mat
                                if ( i .NE. j ) then
                                        off_sum = &
                                        off_sum + mat(i,j)**2
                                end if
                        end do inner2
                end do outer2
        end function mat_offdiag_sqr_norm

        function        jacobi(mat,dim_mat) result(mat_rot)
                integer,intent(in)                      :: dim_mat ! Dimension of matrix
                real(kind = 16),dimension(:,:),allocatable       :: mat ! Cauchy matrix
                real(kind = 16),dimension(:,:),allocatable       :: mat_rot ! Rotation

                real :: mat_max
                real :: s,c,tau
                integer :: row_num,col_num
                real :: t_one,t_two

                allocate( mat_rot(dim_mat,dim_mat) )


                row_num = 1
                col_num = 2
                mat_max = abs( mat(row_num,col_num) )

                outer3 : do i=1,dim_mat
                        inner3 : do j=1,dim_mat
                                if( i .NE. j) then
                                        if( abs( mat(i,j) ) .GT. mat_max ) then
                                                mat_max = abs( mat(i,j) )
                                                row_num = i
                                                col_num = j
                                        end if
                                end if
                        end do inner3
                end do outer3

                tau = ( mat(col_num,col_num) - mat(row_num,row_num) )&
                      /( 2.0 * mat(row_num,col_num) )
                t_one = (-tau) + sqrt(1 + tau**2)
                t_two = (-tau) - sqrt(1 + tau**2)

                if( abs(t_one) .LT. abs(t_two) ) then
                        c = 1.0/sqrt(1+t_one**2)
                        s = c*t_one
                else
                        c = 1.0/sqrt(1+t_two**2)
                        s = c*t_two
                end if

                print *,row_num
               print *,col_num

                outer4 : do i=1,dim_mat
                        inner4 : do j=1,dim_mat
                        if( (i .EQ. row_num) .AND. (j .EQ. row_num) )  then
                                mat_rot(i,j) = c
                        else if( (i .EQ. col_num) .AND. (j .EQ. col_num) ) then
                                mat_rot(i,j) = c
                        else if( (i .EQ. row_num) .AND. (j .EQ. col_num) ) then
                                mat_rot(i,j) = s
                        else if( (i .EQ. col_num) .AND. (j .EQ. row_num) ) then
                                mat_rot(i,j) = (-s)
                        else if( i .EQ. j) then
                                mat_rot(i,j) = 1.0
                        else
                                mat_rot(i,j) = 0.0
                        end if
                        end do inner4
                end do outer4
        end function    jacobi
        function rotate(mat,mat_rot,dim_mat) result(mat_temp)
                integer,intent(in)         :: dim_mat ! Dimension of the matrix
                real(kind = 16),dimension(:,:),allocatable :: mat
                real(kind = 16),dimension(:,:),allocatable :: mat_rot
                real(kind = 16),dimension(:,:),allocatable :: mat_temp

                allocate( mat_temp(dim_mat,dim_mat) )

                mat_temp = matmul(matmul(transpose(mat_rot),mat),mat_rot)


        end function rotate
    	end

File: Jacobi.c
/*Anant Shah
 * Date : 19-12-2017
 **/
#include<stdio.h>
#include<math.h>
#include<stdlib.h>

void mat_gen(double*  mat,int DIM ); /* Generate the Cauchy matrix */
double mat_sqr_norm( double* mat,int DIM ); /* Return the square of the elements of the matrix */
double mat_diag_sqr_norm( double* mat,int DIM ); /* Return the sum of squares of the diagonal elements */
double mat_offdiag_sqr_norm( double* mat,int DIM ); /* Returns the sum of squares of the off diagonal elements */
void jacobi( double* mat,double* mat_rot,int DIM ); /* Creates the rotation matrix */
void ROTATE( double* mat,double* mat_rot,double* mat_mul_temp,double* mat_mul,int DIM); /* Gives the matrix after one jacobi rotation */

int main(int argc,char **argv)
{
        /* Variable Initialization */
        double*         mat; /* Cauchy matrix : A(i,j) =1/i+j */
        int             DIM; /* Dimension of the square matrix */
        double*         mat_mul; /* Multiplication matrix : mat_rot(T)*mat*mat_rot */
        double*         mat_rot; /* Rotation matrix */
        double*         mat_mul_temp; /* Stores the temporary multiplication */
        int             rot_counter = 1; /* Jacobi rotation counter */
        double          sqr_norm; /* Square norm of all elements in the matrix */
        double          diag_sqr_norm; /* Square norm of the diagonal elements */
        double          offdiag_sqr_norm; /* Square norm of the off diagonal elements */
        int             i,j; /* Loop control variables */
        double          sqr_norm_solve; /* Square norm obtained as the sum of diagonal square norm and off diagonal square norm */
        double*         eigenvalues; /* Will store the diagonal elements of the rotated matrix i.e the eigenvalues */


        if(argc!=2)
        {
                printf("error : invalid number of arguments \n");
                exit(0);
        }

        /* Taking arguments from the command line */
        DIM =atoi( argv[1] );

       /* Allocate memory to the matrices */
        mat = (double *)malloc( DIM*DIM*sizeof(double) );
        mat_mul = (double *)malloc( DIM*DIM*sizeof(double) );
        mat_rot = (double *)malloc( DIM*DIM*sizeof(double) );
        mat_mul_temp = (double *)malloc( DIM*DIM*sizeof(double) );
        eigenvalues = (double *)malloc( DIM*sizeof(double) );

        /* Generate the Cauchy matric */
        mat_gen( mat,DIM );

        /* Initialize the temporary multiplication matrices */
        for(i=0;i<DIM;i++)
        {
                for(j=0;j<DIM;j++)
                {
                        mat_mul_temp[i*DIM+j] = 0.0;
                        mat_mul[i*DIM+j] = 0.0;
                }
        }

        sqr_norm = mat_sqr_norm( mat,DIM );
        printf("Square norm of all elements :  \t %.20f \n",sqr_norm);

        while( rot_counter!=0 )
        {
                jacobi( mat,mat_rot,DIM );
                ROTATE( mat,mat_rot,mat_mul_temp,mat_mul,DIM );

                /* Matrix after rotation obtained */
                diag_sqr_norm = mat_diag_sqr_norm( mat,DIM );
                offdiag_sqr_norm = mat_offdiag_sqr_norm( mat,DIM );

                sqr_norm_solve = diag_sqr_norm + offdiag_sqr_norm;
                /* Print the rotation matrix */
                for(i=0;i<DIM;i++)
                {
                        for(j=0;j<DIM;j++)
                        {
                                printf("%.20f ",mat_rot[i*DIM+j]);
                        }
                        printf("\n");
                }

                /* Print the matrix after each rotation */

                if( offdiag_sqr_norm<pow(10,-10) ) {


                        printf("Diagonal sqaure norm : %.20f \n",diag_sqr_norm);
                        printf("Off diagonal square norm : %.20f \n",offdiag_sqr_norm);

                        printf("The obtained square norm is : %.20f \n",sqr_norm_solve);

                        for(i=0;i<DIM;i++)
                        {
                                for(j=0;j<DIM;j++)
                                {
                                        printf("%.20f ",mat[i*DIM+j]);
                                }
                                printf("\n");
                        }
                        break;
                }
                else{
                        printf("Diagonal square norm : \t %.20f \n",diag_sqr_norm);
                        printf("Off Diagonal square norm : \t %.20f \n",offdiag_sqr_norm);

                        printf("The obtained square norm is : \t %.20f \n",sqr_norm_solve);
                        /* Print the matrix */
                        for(i=0;i<DIM;i++){
                                for(j=0;j<DIM;j++){
                                        printf("%.14f ",mat[i*DIM+j]);
                                }
                                printf("\n");
                        }
                }
                rot_counter++;
        }
       /* Assigning the diagonal elements to the <eigenvalues> array */
        for(i=0;i<DIM;i++)
                eigenvalues[i] = mat[i*DIM+i];

        printf("\n \n The eigenvalues of the Cauchy matrix are : \n");
        for(i=0;i<DIM;i++)
                printf("%d. %.14f \n",(i+1),eigenvalues[i]);
}

double mat_sqr_norm(double* mat,int DIM)
{
        /* Variable Initialization */
        int             i,j; /* Loop control variables */
        double          sqr_norm = 0.0; /* Square norm of all the elements */

        for(i=0;i<DIM;i++)
                for(j=0;j<DIM;j++)
                        sqr_norm+=mat[i*DIM+j]*mat[i*DIM+j];

        return sqr_norm;
}

double mat_diag_sqr_norm(double* mat,int DIM)
{
        /* Variable Initialization */
        int             i,j; /* Loop control variables */
        double          diag_sqr_norm = 0.0; /* Square norm of all diagonal elements */

        for(i=0;i<DIM;i++)
                        diag_sqr_norm+=mat[i*DIM+i]*mat[i*DIM+i];

        return diag_sqr_norm;
}

double mat_offdiag_sqr_norm(double* mat,int DIM)
{
        /* Variable Initialization */
        int             i,j; /* Loop control variables */
        double          offdiag_sqr_norm = 0.0; /* Square norm of off diagonal elements */
        for(i=0;i<DIM;i++)
                for(j=0;j<DIM;j++)
                {
                        if(i!=j)
                        {
                                offdiag_sqr_norm+=mat[i*DIM+j]*mat[i*DIM+j];
                        }
                }

        return offdiag_sqr_norm;
}

void ROTATE(double* mat,double* mat_rot,double* mat_mul_temp,double* mat_mul,int DIM)
{
        /* Variable Initialization */
        int             i,j,k; /* Loop initialization variables */

        for(i=0;i<DIM;i++)
                for(j=0;j<DIM;j++)
                        for(k=0;k<DIM;k++)
                                mat_mul_temp[i*DIM+j] += mat_rot[k*DIM+i]*mat[k*DIM+j];

        for(i=0;i<DIM;i++)
                for(j=0;j<DIM;j++)
                        for(k=0;k<DIM;k++)
                                mat_mul[i*DIM+j] += mat_mul_temp[i*DIM+k]*mat_rot[k*DIM+j];

        /* Copy the result back into mul */
        for(i=0;i<DIM;i++)
                for(j=0;j<DIM;j++)
                        mat[i*DIM+j] = mat_mul[i*DIM+j];

        /* Initializa the temporary matrices back to 0 */
        for(i=0;i<DIM;i++)
                for(j=0;j<DIM;j++)
                {
                        mat_mul[i*DIM+j] = 0.0;
                        mat_mul_temp[i*DIM+j] = 0.0;
                }
}

void jacobi(double* mat,double* mat_rot,int DIM)
{
        double          s,c; /* Sin and Cosine to be added to the rotation matrix */
        double          max; /* Maximum absolute values in the cauchy matrix */
        double          tau; /* Value of tau */
        int             i,j; /* Loop control variables */
        int             row_num,col_num; /* Row and column of element with maximum absolute values */
        double          t_one,t_two; /* Two roots of the quadratic; Choose which one is smaller */

        max = fabs( mat[1] );
        row_num = 0;
        col_num = 1;
        /* Find the maximum absolute value in the matrix */
        for(i=0;i<DIM;i++)
                for(j=0;j<DIM;j++)
                {
                        if( ( fabs(mat[i*DIM+j]) > max ) && (i!=j) )
                        {
                                max = fabs(mat[i*DIM + j]);
                                row_num = i;
                                col_num = j;
                        }
                }

        tau = (mat[col_num*DIM+col_num] - mat[row_num*DIM+row_num])/(2.0*mat[row_num*DIM+col_num]);

        t_one = (-tau) + sqrt( 1+tau*tau );
        t_two = (-tau) - sqrt( 1+tau*tau );

        if( fabs( t_one ) < fabs( t_two ) )
        {
                c = 1.0/sqrt( 1+t_one*t_one );
                s = c*t_one;
        }
        else
        {
                c = 1.0/sqrt( 1+t_two*t_two );
                s =  c*t_two;
                printf("%.14f c \n",c);
        }


        /* Rotation mtrix is such that mat(row_num,row_num) = c , mat(col_num,col_num = c) , mat(row_num,col_num) = s, mat(col_num,row_ 
           num) = (-s) , all other diagonal elements = 1, all other off diagonal elements = 0 */
        /* Assign these values into the rotation matrix */
        for(i=0;i<DIM;i++)
        {
                for(j=0;j<DIM;j++)
                {
                        if( (i==row_num) && (j==row_num) )
                                mat_rot[i*DIM+j] = c;
                        else if( (i==col_num) && (j==col_num) )
                                mat_rot[i*DIM+j] = c;
                        else if( (i==row_num) && (j==col_num) )
                                mat_rot[i*DIM+j] = s;
                        else if( (i==col_num) && (j==row_num) )
                                mat_rot[i*DIM+j] = (-s);

                        else if( i==j )
                                mat_rot[i*DIM+j] = 1.0;
                        else
                                mat_rot[i*DIM+j] = 0.0;
                }
        }
}

void mat_gen(double* mat,int DIM)
{
        /* Variable Initialization */
        int             i,j; /*Loop control variables */

        /* Fill in the values of the Cauchy matrix */
        for(i=0;i<DIM;i++)
                for(j=0;j<DIM;j++)
                        mat[i*DIM+j] = (double) 1.0/( (i+1) + (j+1) );
}

